docs/intro/validator.rst

```diff
@@ -0,0 +1,225 @@
+=========
+Validator
+=========
+
+Getting started
+===============
+Validator is an instrument to unify the process of model training and validation.
+There are two main reasons to wrap your model using classes from `Validator` submodule:
+
+- make the structure of the model lifecycle clear,
+- provide API to automized model training and validation.
+
+All you need is:
+
+- define class inherited from `Validator`,
+- configure it by `yaml` file,
+- execute the following code: ::
+
+    from model_api import MyValidator
+
+    val = MyValidator('validator.yaml')
+    val.run()
+    print(val.metrics)
+
+Basic example
+=============
+
+model_api.py
+------------
+::
+
+    import ...
+
+    class PascalValidator(Validator):
+        def __init__(self, *args, **kwargs):
+            super().__init__(*args, **kwargs)
+            self.model_config = ...
+            self.ds = PascalSegmentation(bar=True)
+
+        def train_loader(self, **kwargs):
+            return self.ds.train
+
+        def test_loader(self, **kwargs):
+            return self.ds.test
+
+        def train(self, ds):
+            train_ppl = ... << ds
+            train_ppl.run()
+            self.model = train_ppl.get_model_by_name('model')
+
+        def inference(self, ds):
+            model = self.model
+            test_ppl = ... << ds
+            test_ppl.run()
+            targets = np.array(test_ppl.v('targets'))
+            predictions = np.array(test_ppl.v('predictions'))
+            return targets, predictions
+
+
+validator.yaml
+--------------
+::
+
+    train:
+    test:
+        metrics: accuracy
+
+
+`val.metrics`
+-------------
+
+::
+
+    {'accuracy': 0.0092724609375}
+
+Validator class
+===============
+
+`Validator` child-class can implement the following methods:
+
+`train_loader(self, path=None, **kwargs)`
+-----------------------------------------
+
+`path` and `kwargs` are from config `<train>`. In the example above we have empty value for `train` key, therefore `path=None` and `kwargs={}`. Let's define config:
+
+::
+
+    train:
+        dataset: /path/to/dataset
+
+
+In that case `path='/path/to/dataset'`. You also can define multiple parameters of `train_loader`: ::
+
+    train:
+        dataset:
+        path: /path/to/dataset
+        format: 'png'
+
+Now `path='/path/to/dataset', kwargs={format: 'png'}`.
+The output of the function will be used as `train_dataset` argument of `train` method. By default, it returns `path`.
+
+`test_loader(self, path=None, **kwargs)`
+----------------------------------------
+
+The same as `train_loader` but for `test`.
+
+`load_model(self, path=None, **kwargs)`
+---------------------------------------
+
+Loader for pretrained model. Let's make example above more complex::
+
+    train:
```

Comment:
Why is train here if pretrained is used?

Reply:
"Note that when you define `pretrained` key in your config, train section will be skipped."

Just for fun

Reply:
I've solve to change it:
```
        if 'train' in self.config and 'pretrained' in self.config:
            raise ValueError("Both 'train' and 'pretrained' was founded.")
```

Reply:
Hmm. If we train a pretrained model, should we define both pretrained and train or load model implicitly within train?
I believe both is a better option.

Reply:
You describe a special case but you can implement it using train kwargs. To load and train model you can write the method `train` in the following way:
```
def train(dataset, path_to_model, **kwargs):
    model = load_model(path_to_model)
    ppl = (dataset.p
               ...)
    ...
```
In yaml you can write

```
     train:
        path_to_model: /path/to/model
```

I don't think that we need to implement that custom logic as base functional.